### PR TITLE
chore: Added provision for specific basic auth creds for actuator endpoints

### PR DESCRIPTION
**Pull request by nidhi-nair**
## Description
Adds provision to use custom basic authentication creds for actuator endpoints. If none are provided, we would still need to use basic auth header with empty creds.

This PR requires us to make the relevant changes on cloud (add env vars and change Grafana settings) before it is merged into any of the branches.

Fixes https://github.com/appsmithorg/appsmith-ee/issues/1441

>
#### Type of change
- New feature (non-breaking change which adds functionality)
>
>
#### How Has This Been Tested?
- [x] Manual
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java [hunk 1]
```diff
@@ -176,6 +192,19 @@ public SecurityWebFilterChain securityWebFilterChain(ServerHttpSecurity http) {
...
+    private ReactiveAuthenticationManager getInternalAuthenticationManager() {
+        return authentication -> {
+            String name = authentication.getName();
+            String password = authentication.getCredentials().toString();
+
+            if (INTERNAL_USER_NAME.equals(name) && INTERNAL_PASSWORD.equals(password)) {
+                return Mono.just(UsernamePasswordAuthenticationToken.authenticated(name, password, new ArrayList<>()));
+            } else {
+                return Mono.just(UsernamePasswordAuthenticationToken.unauthenticated(name, password));
+            }
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:204]: Can we change this to keep backwards-compatibility? If username and password aren't set, can we allow this API to be accessible as in the past? Or are we expecting sensitive data to now go out of this endpoint?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java [hunk 2]
```diff
@@ -85,6 +91,12 @@ public class SecurityConfig {
     @Autowired
     private RedirectHelper redirectHelper;
 
+    @Value("${appsmith.internal.username}")
+    private String INTERNAL_USER_NAME;
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:95]: Minor, but this'll be uniform with the env variable.

```suggestion
    private String INTERNAL_USERNAME;
```

**Review by sharat87 — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java (hunk 1 above)
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:204]: @sharat87 Shall I set another environment variable for this that determines whether auth needs to be used? Probably more foolproof than checking for blank fields.

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:204]: Oh nah that's too much for this I think. In fact, I was going to suggest removing username, and just have password. Basic auth can work with just password. One configuration is more that enough for this.

Checking for blank is fine enough, I think. If blank, allow them, if not, require authentication.

Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java]: Okay so Spring does not have an `isBlank` check on conditional properties. I've set up a separate prop for the config right now. Do you think we should make a custom annotation for the check instead? I'm a little divided between maintainability and configurability.

File: app/server/appsmith-server/src/main/resources/application.properties [hunk 3]
```diff
@@ -110,3 +110,7 @@ appsmith.admin.envfile=${APPSMITH_ENVFILE_PATH:/appsmith-stacks/configuration/do
 
 # RTS port
 appsmith.rts.port=${APPSMITH_RTS_PORT:8091}
+
+appsmith.internal.password=${APPSMITH_INTERNAL_PASSWORD:}
+appsmith.internal.auth.enabled=${APPSMITH_INTERNAL_AUTH_ENABLED:false}
+#random.prop=false
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/resources/application.properties:116]: Not random enough I think. :P

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java [hunk 4]
```diff
@@ -176,6 +210,14 @@ public SecurityWebFilterChain securityWebFilterChain(ServerHttpSecurity http) {
                 .build();
     }
 
+    private boolean isAuthorizedToAccessInternal(String password) {
+        if (StringUtils.isNotEmpty(INTERNAL_PASSWORD)) {
+            return INTERNAL_PASSWORD.equals(password);
+        } else {
+            return true;
+        }
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:218]: Feel free to not agree with this suggestion. 🙂

```suggestion
        // Either configured password is empty, or it's equal to what we received.
        return StringUtils.isEmpty(INTERNAL_PASSWORD) || INTERNAL_PASSWORD.equals(password);
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java [hunk 5]
```diff
@@ -109,6 +120,29 @@ public ForwardedHeaderTransformer forwardedHeaderTransformer() {
         return new ForwardedHeaderTransformer();
     }
 
+    @Order(Ordered.HIGHEST_PRECEDENCE)
+    @Bean
+    @ConditionalOnProperty(name = "appsmith.internal.auth.enabled")
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/SecurityConfig.java:125]: Can we use just the password for this? I see we have the implementation where, if password is empty, then auth is always accepted anyway, right? I think we can use `ConditionalOnExpression` to make this conditional on whether `password` is non-empty. One less configuration to care about. What say?

**Review by sharat87 — CHANGES_REQUESTED** [review]

**Review by sharat87 — APPROVED** [review]

**Pull request merged**